bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableSessionThreadPools.java

```diff
@@ -0,0 +1,171 @@
+      BigtableSessionSharedThreadPools shared = BigtableSessionSharedThreadPools.getInstance();
+      if (batchPool == null) {
+        incrementSharedThreadPoolCounter(shared, decrementFromSharedThreadPools);
+        decrementFromSharedThreadPools = true;
+        batchPool = shared.getBatchThreadPool();
+        terminateBatchPool = false;
+      }
+      if (elg == null) {
+        incrementSharedThreadPoolCounter(shared, decrementFromSharedThreadPools);
+        decrementFromSharedThreadPools = true;
+        elg = shared.getElg();
+        terminateElg = false;
+      }
+      if (scheduledRetries == null) {
+        incrementSharedThreadPoolCounter(shared, decrementFromSharedThreadPools);
+        decrementFromSharedThreadPools = true;
+        scheduledRetries = shared.getRetryExecutor();
+        terminateScheduledRetries = false;
+      }
+      return new BigtableSessionThreadPools(batchPool, elg, scheduledRetries,
+          terminateBatchPool, terminateElg, terminateScheduledRetries,
+          decrementFromSharedThreadPools);
+    }
+
+    private void incrementSharedThreadPoolCounter(BigtableSessionSharedThreadPools shared, boolean previouslyIncremented) {
+      if (!previouslyIncremented) {
+        shared.incrementCounter();
+      }
+    }
+  }
+
+  private final ExecutorService batchPool;
+  private final EventLoopGroup elg;
+  private final ScheduledExecutorService scheduledRetries;
+
+  private final boolean terminateBatchPool;
+  private final boolean terminateElg;
+  private final boolean terminateScheduledRetries;
+
+  private final boolean decrementFromSharedThreadPools;
+
+  /**
+   * 
+   * @param batchPool Used for pre and post RPC operations.
+   * @param elg Used by nio
+   * @param scheduledRetries Used for exception handling and retries
+   * @param terminateBatchPool true if the batchPool should be shut down on close.
+   * @param terminateElg true if the elg should be shut down on close.
+   * @param terminateScheduledRetries true if the scheduledRetries should be shut down on close.
+   * @param decrementFromSharedThreadPools true if {@link
+   * BigtableSessionSharedThreadPools#decrementCounter()} should be called on close.
+   */
+  public BigtableSessionThreadPools(
+      ExecutorService batchPool, EventLoopGroup elg, ScheduledExecutorService scheduledRetries, 
+      boolean terminateBatchPool, boolean terminateElg, boolean terminateScheduledRetries, 
+      boolean decrementFromSharedThreadPools) {
+    this.batchPool = batchPool;
+    this.elg = elg;
+    this.scheduledRetries = scheduledRetries;
+    this.terminateBatchPool = terminateBatchPool;
+    this.terminateElg = terminateElg;
+    this.terminateScheduledRetries = terminateScheduledRetries;
+    this.decrementFromSharedThreadPools = decrementFromSharedThreadPools;
+  }
+
+  public ExecutorService getBatchPool() {
+    return batchPool;
+  }
+
+  public EventLoopGroup getElg() {
+    return elg;
```

Comment:
Why not leave null in elg if its not specified in the constructor. And in the accessor return this.elg or shared.getElg(). This way you dont need the bools, you just close this.elg if its not null.